include/deal.II/non_matching/fe_values.h

```diff
@@ -0,0 +1,360 @@
+   *  {
+   *    fe_values.reinit(cell);
+   *
+   *    const boost::optional<const FEValues<dim> &> fe_values_inside =
+   *      fe_values.get_inside_fe_values();
+   *
+   *    if (fe_values_inside)
+   *      {
+   *        // Assemble locally
+   *        for (const unsigned int q_index :
+   *             fe_values_inside->quadrature_point_indices())
+   *          {
+   *            // ...
+   *          }
+   *      }
+   *  }
+   * @endcode
+   *
+   * Of course, it is somewhat expensive to generate the immersed quadrature
+   * rules and create FEValues objects with the generated quadratures. To reduce
+   * the amount of work, the reinit() function of this class uses the
+   * MeshClassifier passed to the constructor to check how the incoming cell
+   * relates to the level set function. It only generates the immersed
+   * quadrature rules if the cell is intersected. If the cell is completely
+   * inside or outside it returns a cached FEValues object created with a
+   * quadrature over the reference cell: $[0, 1]^{dim}$.
+   */
+  template <int dim>
+  class FEValues
+  {
+  public:
+    using AdditionalData = typename QuadratureGenerator<dim>::AdditionalData;
+
+    /**
+     * Constructor.
+     *
+     * @param mapping_collection Collection of Mappings to be used.
+     * @param fe_collection Collection of FiniteElements to be used.
+     * @param q_collection Collection of Quadrature rules over $[0, 1]^{dim}$
+     * that should be used when a cell isn't intersected and we don't need to
+     * generate immersed quadrature rules.
+     * @param q_collection_1D Collection of 1-dimensional quadrature rules used
+     * to generate the immersed quadrature rules. See the QuadratureGenerator
+     * class.
+     * @param mesh_classifier Object used to determine when the immersed
+     * quadrature rules need to be generated.
+     * @param region_update_flags Struct storing UpdateFlags for the
+     * inside/outside/surface region of the cell.
+     * @param dof_handler The DoFHandler associated with the discrete level set
+     * function.
+     * @param level_set The degrees of freedom of the discrete level set function.
+     * @param additional_data Additional data passed to the QuadratureGenerator
+     * class.
+     *
+     * @note Pointers to @p mapping_collection, @p fe_collection,
+     * @p mesh_classifier, @p dof_handler, and @p level_set are stored
+     * internally, so these need to have a longer life span than the instance of
+     * this class.
+     *
+     * @note Only the case when @p mapping_collection contains MappingCartesian
+     * is presently implemented.
+     */
+    template <class VECTOR>
+    FEValues(const hp::MappingCollection<dim> &mapping_collection,
+             const hp::FECollection<dim> &     fe_collection,
+             const hp::QCollection<dim> &      q_collection,
+             const hp::QCollection<1> &        q_collection_1D,
```

Comment:
As stated in #12757, I would like to see a version without the collections. Also, I think the 1D quadrature can be simply generated from the higher-dimensional one.

Reply:
I've added an easier-to-use constructor. It's not exactly what you asked for, but I think it makes sense.